v2/tools/generator/internal/codegen/pipeline/assemble_oneof.go

```diff
@@ -0,0 +1,418 @@
+	}
+
+	return result
+}
+
+// assembleOneOfs is called to build up the oneOf types.
+// We return a TypeDefinitionSet containing the new types.
+func (oa *oneOfAssembler) assembleOneOfs() (astmodel.TypeDefinitionSet, error) {
+	// Find all OneOfs and work out the modifications we need to make to each one
+	var errs []error
+	for tn := range oa.defs {
+		if _, ok := oa.asOneOf(tn); ok {
+			err := oa.assemble(tn)
+			if err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
+
+	// If we encountered any errors, bail out
+	err := kerrors.NewAggregate(errs)
+	if err != nil {
+		return nil, err
+	}
+
+	// Assemble our results by applying our accumulated modifications
+	result := make(astmodel.TypeDefinitionSet)
+	for tn, oneOf := range oa.updates {
+		def := oa.defs.MustGetDefinition(tn)
+		result.Add(def.WithType(oneOf))
+	}
+
+	return result, nil
+}
+
+// assemble is called to build up a single oneOf type.
+func (oa *oneOfAssembler) assemble(name astmodel.TypeName) error {
+	for _, tn := range oa.findParentsOf(name) {
+		err := oa.assemblePair(tn, name)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// assemblePair updates a pair of nodes that have a relationship and recursively invokes itself to traverse the entire
+// tree of referenced imports.
+// ancestor is the parent or super-type. It may be a less specialised OneOf, a root OneOf, an Object or an AllOf
+// descendent is the child or subtype. It will be an intermediate or leaf OneOf.
+func (oa *oneOfAssembler) assemblePair(ancestor astmodel.TypeName, descendant astmodel.TypeName) error {
+	// Remove any direct reference to the parent from the leaf
+	err := oa.removeParentReferenceFromLeaf(ancestor, descendant)
+	if err != nil {
+		return errors.Wrapf(err, "removing parent reference %s from leaf %s", ancestor, descendant)
+	}
+
+	if oa.hasDiscriminatorValue(descendant) {
+		// Copy any common properties from the parent to the child
+		err = oa.embedCommonPropertiesInLeaf(ancestor, descendant)
+		if err != nil {
+			return errors.Wrapf(err, "embedding common properties from %s in leaf %s", ancestor, descendant)
+		}
+	}
+
+	if oa.isDiscriminatorRoot(ancestor) {
+		// Strip common properties from the root (we replace these with other properties in a later stage)
+		err := oa.removeCommonPropertiesFromRoot(ancestor)
+		if err != nil {
+			return errors.Wrapf(err, "removing common properties from OneOf root %s", ancestor)
+		}
+
+		// We've found the actual root, add a reference to the leaf
+		err = oa.addLeafReferenceToRoot(ancestor, descendant)
+		if err != nil {
+			return errors.Wrapf(err, "adding leaf reference to %s from root %s", descendant, ancestor)
+		}
+
+		// Ensure the discriminator property exists on the leaf
+		err = oa.addDiscriminatorProperty(descendant, ancestor)
+		if err != nil {
+			return errors.Wrapf(err, "ensuring discriminator property exists on %s", descendant)
+		}
+	}
+
+	// Recursively walk any parent's of the current ancestor
```

Comment:
Just confirming: This is needed because of the possibility of nested sets of these hierarchies? (ick)

Reply:
Yup. And agreed on 'ick'.